### PR TITLE
Basic gradient saliency visualization

### PR DESCRIPTION
**Pull request by Nimanui**
Had to refresh the branch due to some github shenanigans. Should be the same as the previous pull request, but I completed adjustments to let it render and manage batches instead of inputting a dataloader (I left the dataloader stuff if someone wants to do a larger set but perhaps it's redundant). I've also updated the example python notebook accordingly and refactored the tests. Also did some light refactoring to emphasize that this is just a basic pytorch gradient saliency visualization class.

### CODE REVIEW

**Comment by jhnwu3** [comment]
Running unit tests, if it passes, I think it looks pretty good to me. We can always iterate further later. I think at some point, we might need a standardized API for the interpretability module haha since I see a lot of similarities in our approaches here.

**Review by jhnwu3 — APPROVED** [review]
lgtm

File: pyhealth/interpret/methods/basic_gradient.py
Comment (jhnwu3) [inline pyhealth/interpret/methods/basic_gradient.py:1]: Ahh, sorry, I should've reviewed more in detail earlier. Was stuck in a meeting.

Some quick thoughts that I think would make more sense in terms of a refactor here again (sorry to make you do this, but I think it would reduce the complexity of the approach.

1. We shouldn't store the dataloader inside of the class as it's really heavy to have an entire dataloader inside of an interpretability class here. The idea here is we simply call this class's "attribute" or .saliency_map(batch) as we're iterating across the dataloader. This way, technically the user can define whether or not to do it across the entire dataset themselves or decide how they want to do the sampling of interpretation.

2. If possible, the .attribute() or .process_batch() would just return the tensor attributions themselves as I think that would be a pretty straightforward way of letting the user use these attribution features.

3. The reason why we want to return the raw attribution tensor features is because a lot of downstream interpretability methods (like adversarial learning approaches) can often use these attributions in other creative ways.

**Review by jhnwu3 — CHANGES_REQUESTED** [review]

File: pyhealth/interpret/methods/basic_gradient.py
Comment (jhnwu3) [reply pyhealth/interpret/methods/basic_gradient.py:1]: Can we also have it inherit from the new master's BaseInterpreterClass?

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: I could try but this already far more overengineered than the intent of just a basic method that just does a very simple gradient on input image data from a CNN. Would it be possible to come back to that on a later pass and then if this is helpful I can expand it to support StageNet at that point?

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: I also would like to try to use some minor image support with the LRP but if I reference any of this code or the example in it, it becomes dependent on this and a bit of a problem. I'm happy to come back to this later but I'm kind of stuck with this still being open.

Comment (jhnwu3) [reply pyhealth/interpret/methods/basic_gradient.py:1]: Yes exactly, I think we can definitely simplify this class substantially. There are some things I'm a little confused about. Let me know your thoughts.

1. We can remove the input_batch init argument here? Do we need an assertion check here? 
2. It feels like _process_batch and attribute() are redundant here?
3. I also feel like _compute_saliency() isn't called anywhere? Or get_saliency_maps()?
4. I really like the visualize function idea

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
Oh on the topic of documentation, can you add this file reference to: https://github.com/sunlabuiuc/PyHealth/blob/master/docs/api/interpret.rst

File: pyhealth/interpret/methods/basic_gradient.py
Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: Good thoughts, and reasonable questions. 

1. The goal of the input batch is to allow the user to provide a specific set of pictures/raw data that they want to process through the interpretability process. Otherwise I would have to add a large quantity of sorting or other tools to figure out which data they want to use, or go back to the original implementation which just processed all of it (with some options to restrain it a bit for performance). Maybe I'm misunderstanding the base class, but it doesn't feel like we currently have a consistent way to help users decide what data they want to look at (perhaps doubly important with the visualization applictations).
2. You are correct that there is some functional overlaps there, I'll merge them a bit to reduce the redundancy
3. Good call, I think it may have gotten lost in the shuffle, I think the get is being called in the python notebook example but may also be extra.
4. I think I'll try to refactor that into a general class in a future iteration, part of the push to get this through is that you can visualize LRP and it's quite a fun example, but it mildly depends on those functions here. In particular having the base image underneath the saliency map makes for an excellent visual (imo).

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: Ok, found a few more spots, part of the issue is when I devoloped this I figured it would be better to build a dictionary that has the input image, saliency map, and label all kept together. Otherwise, if we only store the saliency map on it's own, it will quickly become impossible to tie it back to the original image for the visualization (and well they are somewhat helpful on their own, I really like the idea of having the original image underneath). So the _compute_saliency function is just setting up that dictionary and constructing it. I agree that the getter is a bit redundant, that was originally because I had it using a lazy initialization where it didn't compute all of the saliency maps until they were specifically asked for. This made more sense when it had the "process the entire inputted dataset" option originally and isn't as helpful now that it is only doing processing by batch anyway. I'll tie the initialize into the main init and just leave a simple getter for testing use if someone wants to grab the saliency dictionary.

Comment (jhnwu3) [reply pyhealth/interpret/methods/basic_gradient.py:1]: 1. So the attribute(**data_batch) should be able to let users pass in any batch and return their attributions here. For reference, you can iterate through a pytorch dataloader and filter for a specific input. I think my concern is that it doesn't make sense for it to be in the init function? When other functions clearly can perform the attribution on any batch the user decides?



4. (4) No worries, we can leave those be. Happy to have them as they are now.

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: 1. At that point then, isn't it just a static function and class at that point then? I'm not sure what you would store from in the class if you don't even have the processing batch on init. I could make it optional for the existing application (which only calls attribute as a helper method), and I'll adjust it so it updates the batch if one is passed with the attribute method. I think that lines it up a bit closer.
2. Turns out it slimmed up real nice, thank you for poking about this. I was able to clear about 100 lines and I'm pretty happy with the results.

Comment (jhnwu3) [reply pyhealth/interpret/methods/basic_gradient.py:1]: 1. The main thing I think (and feel free to disagree with me on this) is just making things standardized where we can assume that every (local) interpretability method should have an attribute() function that can be leveraged in our interpretability metrics here since the metrics assume you have some type of function attribute(), and it's nice to centralize the logic around this abstract interpreter class.

There's also things like not being able to compute gradients on discrete variables (see the DeepLift and IntegratedGradient examples) where you'll need some type of specialized logic here. I would imagine if these gradient-based saliency maps were computed on things like discrete word tokens here. It would require a little more complicated logic here, whether that's storing different types of baseline inputs to compute some difference in activations or just being able to have different modes (i.e SHAP can be computed a variety of different ways, but at the end of the day, they're all estimating the true SHAPley values)

But, honestly, I think object-oriented approaches are just a little cleaner in terms of abstraction and code readability (for me personally) Technically if we were to follow a pythonic style, we could pass around a bunch of functions everywhere, but imo that would be really hard to read.

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: Ok, those changes are in as well, let me know if you find anything else. There is probably still a few minor refactoring pieces I can think of but I think that's all of the major ones. If there are additional examples we want to add I'd be inclined to do that later.

Comment (Nimanui) [reply pyhealth/interpret/methods/basic_gradient.py:1]: And the notebook is up to date with this code, although again I'm not sure if we want to do some more notebooks with different core models. I only picked this one as it was an easy fit, but I think we could add it to the chestxray 14 multiclassifier as well. There are also plenty of fun options for adding visualization tools, again thoughts for later work.

File: docs/api/interpret/pyhealth.interpret.methods.basic_gradient.rst
Comment (jhnwu3) [inline docs/api/interpret/pyhealth.interpret.methods.basic_gradient.rst:1]: For the sake of vibe-coding purposes, can we move these details into the docstrings in the class implementation? Also, it seems the docs are out of date. The nice part about having it specifically in the doc strings (that get rendered by autodocs) is that the LLM will update the doc strings and the logic every iteration so it'll be easy to keep things maintained.

I think the examples/notebook might be out of date too?

Otherwise, I think we're pretty much there. Thank you for taking the time to iterate on this.

Comment (Nimanui) [reply docs/api/interpret/pyhealth.interpret.methods.basic_gradient.rst:1]: The notebook is updated (I've been testing with it for every change in addtion to the unit tests), I'll poke the documentation and let you know, that probably is out of date. I don't mind moving it, the other documentation I just added is up to date.

Comment (Nimanui) [reply docs/api/interpret/pyhealth.interpret.methods.basic_gradient.rst:1]: The rst is updated to resemble the other interpretability ones (sorry I didn't do that sooner) and I added the gist of what it had to the doc string in the code. Excellent call there and thank you for helping me make this much better.
